src/addons/gif.cpp

```diff
@@ -0,0 +1,117 @@
+#include "configs/base64.h"
+#include "addons/gif.h"
+#include <hardware/flash.h>
+#include "FlashPROM.h"
+#include "FlashPROM.cpp"
+#include "storagemanager.h"
+
+extern char __flash_binary_end;
+
+bool GifAddon::available() {
+    const DisplayOptions& displayOptions = Storage::getInstance().getDisplayOptions();
+    const GifOptions& gifOptions = Storage::getInstance().getAddonOptions().gifOptions;
+    return displayOptions.enabled && gifOptions.enabled;
+};
+
+void GifAddon::setup() {
+    setCurrentFrameFor(0);
+};
+
+bool GifAddon::setupForWrite(size_t numberOfBytes)
+{
+    if (get_core_num() != 0)
+        return false;
+
+    const uint32_t numberOfFrames = numberOfBytes / sizeof(SplashFrame);
+
+    const size_t framesMaxAddress = EEPROM_ADDRESS_START - XIP_BASE - GIF_EEPROM_DISTANCE;
+    const size_t alignedFramesMaxAddress = (framesMaxAddress / FLASH_SECTOR_SIZE == 0)
+        ? framesMaxAddress
+        : framesMaxAddress - framesMaxAddress % FLASH_SECTOR_SIZE;
+
+    const size_t framesMinAddress = (size_t) &__flash_binary_end  + GIF_CODE_DISTANCE - XIP_BASE;
+    const size_t alignedFramesMinAddress = (framesMinAddress / FLASH_SECTOR_SIZE == 0)
+        ? framesMinAddress
+        : framesMinAddress + FLASH_SECTOR_SIZE - framesMinAddress % FLASH_SECTOR_SIZE;
+
+    const size_t availableMemory = alignedFramesMaxAddress - alignedFramesMinAddress;
+
+    if (availableMemory <= 0) return false;
+
+    const size_t framesSize = sizeof(SplashFrame) * numberOfFrames;
+    const size_t numberOfSectors = (framesSize % FLASH_SECTOR_SIZE == 0)
+        ? framesSize / FLASH_SECTOR_SIZE 
+        : framesSize / FLASH_SECTOR_SIZE + 1;
+
+    const size_t necessaryMemory = numberOfSectors * FLASH_SECTOR_SIZE;
+
+    if (necessaryMemory > availableMemory) return false;
+
+    const size_t startingAddress = alignedFramesMaxAddress - necessaryMemory;
+
+    while (is_spin_locked(flashLock));
+
+    multicore_lockout_start_blocking();
```

Comment:
is this thread safe for when USB is processing? or will this only happen during the web config setup

Reply:
I'm not keen to the details but the functions are the same as the ones that are being used on the EEPROM simulation library used by the project